core/Pos/Core/Genesis/Types.hs

```diff
@@ -57,20 +51,31 @@ getTotalStake ExponentialStakes = mkCoin . sum $
     let g 0 = []
         g n = n : g (n `div` 2)
     in g 5000
-getTotalStake (ExplicitStakes balances) = unsafeIntegerToCoin $
-    sumCoins $ fst <$> balances
-getTotalStake (CombinedStakes st1 st2) =
-    getTotalStake st1 `unsafeAddCoin` getTotalStake st2
+getTotalStake (CustomStakes balances) =
+    unsafeIntegerToCoin $ sumCoins balances
 
--- | Hardcoded genesis data
-data GenesisCoreData = GenesisCoreData
-    { gcdAddresses         :: [Address]
-    , gcdDistribution      :: StakeDistribution
-    , gcdBootstrapBalances :: !(HashMap StakeholderId Coin)
-    }
-    deriving (Show, Eq)
+-- | Distributions accompained by related addresses set (what to
+-- distribute and how).
+type AddrDistribution = ([Address], StakeDistribution)
+
+-- | Hardcoded genesis data to generate utxo from.
+data GenesisCoreData = UnsafeGenesisCoreData
+    { gcdAddrDistribution      :: !([AddrDistribution])
+      -- ^ Address distribution. Determines utxo without boot
+      -- stakeholders distribution (addresses and coins).
+    , gcdBootstrapStakeholders :: !(HashSet StakeholderId)
+      -- ^ Bootstrap era stakeholders.
+    } deriving (Show, Eq)
 
-instance Monoid GenesisCoreData where
-    mempty = GenesisCoreData mempty mempty mempty
-    (GenesisCoreData addrsA distA bbsA) `mappend` (GenesisCoreData addrsB distB bbsB) =
-        GenesisCoreData (addrsA <> addrsB) (distA <> distB) (bbsA <> bbsB)
+-- | Safe constructor for 'GenesisCoreData'. Throws error if something
+-- goes wrong.
+mkGenesisCoreData ::
+       [AddrDistribution]
+    -> HashSet StakeholderId
+    -> Either String GenesisCoreData
+mkGenesisCoreData distribution bootStakeholders = do
+    -- TODO CSL-1205 add checks
```

Comment:
??
What's the purpose of referring to **this** issue? Should it still be considered WIP?
Or do you want to do it in another PR?
If I merge this PR and we close CSL-1205, there will be a TODO referring to the closed issue which is bad.